### PR TITLE
feat(pattern): add a tooltip for each fields that shows statistics about the field.

### PR DESCRIPTION
**Pull request by cyriltovena**
This adds a clickable tooltip for each fields in found patterns.
The tooltip shows statistics about the last 2k found samples. 

This PR also refactor Patterns component into their own folder.

image

### CODE REVIEW

**Comment by matyax** [comment]
Hey @cyriltovena . Are you expecting our review here? Just asking because there are no assigned reviewers, and I missed this one.

**Comment by cyriltovena** [comment]
Please yes ! Don't hesitate to let me know what I can improve, I'm still learning frontend.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx:14]: bit of a nit: but I think we're trying to stick to the css object styles in emotion instead of the plain css/string type: https://emotion.sh/docs/object-styles

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:68]: I don't think we need the cx function, cx is used for combining the styles from multiple classes. `className={styles.logsStatsRowLabel}` should work fine. Also cx has some performance impact so it should be avoided when possible

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:82]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/displayName

I'm not super familiar with how this will work since it's non-standard, I guess there isn't any user facing functionality that uses it, but I'd rather err on the side of caution and remove it? @matyax what do you think?

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:46]: Need to catch errors and display error state. Currently failures load forever.
image

Comment (gtk-grafana) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:46]: Request that's failing from above:

Expr:
```
{service_name=`grafana/grafana`} |> `logger=sqlstore.metrics traceID=<_> t=<_>:<_>:<_>.<_> level=debug msg="query finished" status=success elapsedtime=<_>.<_> sql="SELECT `id`, `org_id`, `title`, `condition`, `data`, `updated`, `interval_seconds`, `version`, `uid`, `namespace_uid`, `dashboard_uid`, `panel_id`, `rule_group`, `rule_group_idx`, `record`, `no_data_state`, `exec_err_state`, `for`, `annotations`, `labels`, `is_paused`, `notification_settings` FROM `alert_rule` WHERE `rule_group` IN (?)" error=null` | pattern `logger=sqlstore.metrics traceID=<field_1> t=<field_2>:<field_3>:<field_4>.<field_5> level=debug msg="query finished" status=success elapsedtime=<field_6>.<field_7> sql="SELECT `id`, `org_id`, `title`, `condition`, `data`, `updated`, `interval_seconds`, `version`, `uid`, `namespace_uid`, `dashboard_uid`, `panel_id`, `rule_group`, `rule_group_idx`, `record`, `no_data_state`, `exec_err_state`, `for`, `annotations`, `labels`, `is_paused`, `notification_settings` FROM `alert_rule` WHERE `rule_group` IN (?)" error=null` | keep field_1 ,field_2 ,field_3 ,field_4 ,field_5 ,field_6 ,field_7 | line_format ""
```

Full request:
```
ds_type: loki
requestId: 1_1
{"queries":[{"refId":"A","queryType":"range","editorMode":"code","supportingQueryType":"grafana-lokiexplore-app","maxLines":2000,"expr":"{service_name=`grafana/grafana`} |> `logger=sqlstore.metrics traceID=<_> t=<_>:<_>:<_>.<_> level=debug msg=\"query finished\" status=success elapsedtime=<_>.<_> sql=\"SELECT `id`, `org_id`, `title`, `condition`, `data`, `updated`, `interval_seconds`, `version`, `uid`, `namespace_uid`, `dashboard_uid`, `panel_id`, `rule_group`, `rule_group_idx`, `record`, `no_data_state`, `exec_err_state`, `for`, `annotations`, `labels`, `is_paused`, `notification_settings` FROM `alert_rule` WHERE `rule_group` IN (?)\" error=null` | pattern `logger=sqlstore.metrics traceID=<field_1> t=<field_2>:<field_3>:<field_4>.<field_5> level=debug msg=\"query finished\" status=success elapsedtime=<field_6>.<field_7> sql=\"SELECT `id`, `org_id`, `title`, `condition`, `data`, `updated`, `interval_seconds`, `version`, `uid`, `namespace_uid`, `dashboard_uid`, `panel_id`, `rule_group`, `rule_group_idx`, `record`, `no_data_state`, `exec_err_state`, `for`, `annotations`, `labels`, `is_paused`, `notification_settings` FROM `alert_rule` WHERE `rule_group` IN (?)\" error=null` | keep field_1 ,field_2 ,field_3 ,field_4 ,field_5 ,field_6 ,field_7 | line_format \"\"","legendFormat":"","datasource":{"type":"loki","uid":"e80c5725-e146-4f8b-bdf7-82a037b21cbc"},"datasourceId":350,"intervalMs":0}],"from":"1719415107085","to":"1719416007086"}
```

Pattern:
```
logger=sqlstore.metrics traceID=<_> t=<_>:<_>:<_>.<_> level=debug msg="query finished" status=success elapsedtime=<_>.<_> sql="SELECT `id`, `org_id`, `title`, `condition`, `data`, `updated`, `interval_seconds`, `version`, `uid`, `namespace_uid`, `dashboard_uid`, `panel_id`, `rule_group`, `rule_group_idx`, `record`, `no_data_state`, `exec_err_state`, `for`, `annotations`, `labels`, `is_paused`, `notification_settings` FROM `alert_rule` WHERE `rule_group` IN (?)" error=null
```

Response:
```
{
    "results": {
        "A": {
            "error": "parse error at line 1, col 173: syntax error: unexpected IDENTIFIER",
            "errorSource": "downstream",
            "status": 500
        }
    }
}
```

Found in Grafana logging dev

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:82]: This is not needed here at all. It's used when you wrap your component in, for example, `React.memo(() => {})`.
Tools like the React developer tools use this display name to show you which component has been rendered in the tree.

E.g. https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/DateTimePickers/DatePicker/DatePicker.tsx#L40

**Comment by gtk-grafana** [comment]
0 results found:
image
image
Ideally we could not make this linkable, but we likely don't know beforehand?

**Comment by gtk-grafana** [comment]
Duration split up by the decimal:
image

Again not sure we can do anything here, but knowing the distribution of decimals is not helpful to the user.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:10]: theme.spacing(1) is 4. This could give you a similar number, but I would recommend to always use whole numbers or quarters for spacing. E.g. spacing(1), spacing(0.25), spacing(3.5).


```suggestion
    margin: ${theme.spacing(1.15)}px 0;
```

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:21]:
```suggestion
  const [stats, setStats] = useState<LogLabelStatsModel[][]>([]);
```

This would simplify the code a bit, as you don't need to check if `stats` is undefined.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx:14]: Some example values can get really long, pushing the content off the page
image

Setting a max height and overflow: auto will allow the modal to scroll when the content doesn't fit. 

e.g.
```
      overflow: auto;
      max-height: 80vh;
```
80vh might be too tall, something we should run by @zizzpudding

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (gtk-grafana) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:69]: The title tag looks a little funny and doesn't really provide any additional benefit unless we are truncating this text or displaying some other information to the user

image

**Comment by gtk-grafana** [comment]
Another design question, but not something that needs to be addressed in this PR:

image
image
image
image


Should other be given a special treatment, or hardcoded placement, instead of always sorting by percentage?
@zizzpudding

**Comment by gtk-grafana** [comment]
Keyboard nav works, which is awesome, and it works as expected, except that opening a modal with keyboard doesn't close existing one like the mouse UX

image

**Review by gtk-grafana — DISMISSED** [review]
Overall really great work on this, and this is a very powerful feature that really helps explain to the user what the patterns are and how they work, but there's a few bugs/code cleanup we need to look into and we should review with @zizzpudding before release.

@cyriltovena Let us know if you want to hand any of this off if you have other priorities.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (matyax) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:69]: Very clever 👍

**Comment by matyax** [comment]
Love the feature!

I'm curious if we can do a better job explaining this distribution, because it can be challenging to understand what this means.

imagen

Maybe we can just let the user know that "this just has many many values".

**Comment by gtk-grafana** [comment]
> Maybe we can just let the user know that "this just has many many values".

I like that it still shows example values, but I agree that there should be a better way to show the user that the cardinality of these values is maxed out

**Comment by cyriltovena** [comment]
feedback from @matryer should we just show % without the bar ?

**Comment by cyriltovena** [comment]
> Duration split up by the decimal: image image image
> 
> Again not sure we can do anything here, but knowing the distribution of decimals is not helpful to the user.

We're working on this on the backend side. Spotted that one too recently.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx]: This is done.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx]: Removed.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx]: Done.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStatsRow.tsx:73]: Removed.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:47]: Errors are now being handled. I will adjust styles in general in further commits.

imagen

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx
Comment (matyax) [reply src/Components/ServiceScene/Breakdowns/Patterns/PatternFieldLabelStats.tsx]: Done, set to 40vh.

Scrolla

**Comment by matyax** [comment]
Most, or all, comments and issues have been addressed.

File: src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx
Comment (svennergr) [inline src/Components/ServiceScene/Breakdowns/Patterns/PatternNameLabel.tsx:152]: smart 👍

**Review by svennergr — APPROVED** [review]
LGTM. Nice feature. I think we can iterate after merging this on the UX, but let's gooo

**Comment by matyax** [comment]
BTW, can't merge this one because of the requested changes.

**Pull request merged**
